### PR TITLE
Fix eth_estimateGas 0 amounts

### PR DESCRIPTION
**Pull request by DarthMike**

### CODE REVIEW

File: web3swift/src/Client/EthereumClient.swift [hunk 1]
```diff
@@ -177,13 +177,18 @@ public class EthereumClient: EthereumClientProtocol {
...
+                
+                if let gas = gas.map(jsonRPCAmount) {
                     try nested.encode(gas, forKey: .gas)
                 }
-                if let gasPrice = gasPrice {
+                if let gasPrice = gasPrice.map(jsonRPCAmount) {
                     try nested.encode(gasPrice, forKey: .gasPrice)
                 }
-                if let value = value {
+                if let value = gas.map(jsonRPCAmount) {
```
Comment (th3m477) [inline web3swift/src/Client/EthereumClient.swift:191]: value?

File: web3swift/src/Contract/Statically Typed/ABIFunction.swift [hunk 2]
```diff
@@ -16,17 +16,16 @@ public protocol ABIFunction {
...
     var from: EthereumAddress? { get }
     func encode(to encoder: ABIFunctionEncoder) throws
-    func transaction() throws -> EthereumTransaction
 }
 
 public protocol ABIResponse: ABITupleDecodable {}
 
 extension ABIFunction {
-    public func transaction() throws -> EthereumTransaction {
+    public func transaction(gasPrice: BigUInt? = nil, gasLimit: BigUInt? = nil) throws -> EthereumTransaction {
```
Comment (th3m477) [inline web3swift/src/Contract/Statically Typed/ABIFunction.swift:24]: y?

**Review by th3m477 — CHANGES_REQUESTED** [review]

File: web3swift/src/Contract/Statically Typed/ABIFunction.swift (hunk 2 above)
Comment (DarthMike) [reply web3swift/src/Contract/Statically Typed/ABIFunction.swift:24]: Needed to override gasPrice and/or gasLimit in `ABIFunction`, i.e. estimating with no refunds

File: web3swift/src/Client/EthereumClient.swift (hunk 1 above)
Comment (DarthMike) [reply web3swift/src/Client/EthereumClient.swift:191]: good spot , fixed

File: web3swift/src/Contract/Statically Typed/ABIFunction.swift (hunk 2 above)
Comment (th3m477) [reply web3swift/src/Contract/Statically Typed/ABIFunction.swift:24]: Could just pass in a different ABIFunction though/make mutable; might seem more ambiguous down at this level? Approved PR so your call.

**Review by th3m477 — APPROVED** [review]

File: web3swift/src/Contract/Statically Typed/ABIFunction.swift (hunk 2 above)
Comment (DarthMike) [reply web3swift/src/Contract/Statically Typed/ABIFunction.swift:24]: Needs refactor everywhere + change of API. `ABIFunction` protocol allows for read-only values. We should also differentiate with calls, will change later
👍

**Pull request merged**